OMEdit/OMEditGUI/Editors/BaseEditor.cpp

```diff
@@ -1442,16 +1478,45 @@ QCompleter *PlainTextEdit::completer()
   return mpCompleter;
 }
 
-/* insert the keywords from the completer popup */
-void PlainTextEdit::insertCompletion(const QString& completion)
+/*!
+ * PlainTextEdit::insertCompletionItem
+ * insert the completerItems from the completer popup
+ */
+void PlainTextEdit::insertCompletionItem(const QModelIndex &index)
 {
+  QVariant value = index.data(Qt::UserRole);
+  CompleterItem completerItem = qvariant_cast<CompleterItem>(value);
+  QString selectiontext = completerItem.select;
+  QStringList completionlength = completerItem.value.split("\n");
   QTextCursor cursor = textCursor();
-  int extra = completion.length() - mpCompleter->completionPrefix().length();
-  //cursor.movePosition(QTextCursor::Left);
+  int extra = completionlength[0].length() - mpCompleter->completionPrefix().length();
   cursor.movePosition(QTextCursor::EndOfWord);
-  cursor.insertText(completion.right(extra));
+  cursor.insertText(completionlength[0].right(extra));
+  // store the cursor position to be used for selecting text when inserting code snippets
+  int currentpos = cursor.position();
+  int startpos = currentpos-completionlength[0].length();
   setTextCursor(cursor);
-}
+  // To insert CodeSnippets
+  if (completionlength.length()>1)
```

Comment:
Here i think if we check that it is keyword, or type then we have to write the same code thrice for each cases